runtime/doc/lsp.txt

```diff
@@ -886,10 +888,11 @@ start_client({config})                                *vim.lsp.start_client()*
       • {config}  (`vim.lsp.ClientConfig`) Configuration for the server. See
                   |vim.lsp.ClientConfig|.
 
-    Return: ~
+    Return (multiple): ~
         (`integer?`) client_id |vim.lsp.get_client_by_id()| Note: client may
         not be fully initialized. Use `on_init` to do any actions once the
         client has been initialized.
+        (`string?`) Error message, if any
```

Comment:
Ideally we should follow the `pcall` convention of `boolean, string|result` but it's probably too breaking to do that here?

Reply:
Not sure if it's too breaking, but I'd like to point out that `vim.uv` uses this pattern of `result|nil, errormsg|nil` for most of its functions. So in a way the pattern here wouldn't be new. Probably rather the opposite, there are fewer cases where `boolean, string|result` is used.

Reply:
The Lua stdlib does this too, e.g. `io.open` returns `file|nil, nil|error`.

Reply:
There is also an advantage to the `result|nil, errormsg|nil` pattern:
If you do:

```
local result, err = foobar()
if err then
    ...
elseif result then
    ...
end
```

luals can correctly infer the type of `result`.
With `bool, result|err` a `if ok` won't tell it if the second return value has the type of the error, or the type of the result.

Reply:
interesting. let's document this pattern in `:help dev-lua` "interface conventions" section.

I should probably rework https://github.com/neovim/neovim/pull/28502 ?